src/api/common/WorkerProxy.ts

```diff
@@ -44,6 +44,39 @@ export function exposeLocal<T, IncomingRequestType>(impls: T): (message: Request
 	}
 }
 
+export type FacadeImpls = {
+	[facade: string]: any
+}
+
+export type DelayedImpls<IMPLS extends FacadeImpls> = {
+	[Property in keyof IMPLS]: () => Promise<IMPLS[Property]>
+}
+
+/**
+ * Generate a handler which will delegate to {@param impls}.
+ * Attention! Make sure that the *only* fields on T are functions that resolve to facades. Every facade method must return promise or Bad Things will happen.
+ * You should specify T explicitly to avoid mistakes.
+ */
+export function exposeLocalDelayed<T extends DelayedImpls<FacadeImpls>, IncomingRequestType>(
+	impls: T,
+): (message: Request<IncomingRequestType>) => Promise<any> {
+	return async (message: Request<IncomingRequestType>) => {
+		const [facade, fn, args] = message.args
+		const init = downcast(impls)[facade]
+
+		if (init == null) {
+			throw new ProgrammingError(`Facade is not exposed: ${facade}.${fn} (exposeLocal)`)
+		}
+
+		const impl = await init()
+		if (impl == null) {
+			throw new ProgrammingError(`Facade is not lazy: ${facade}.${fn} (exposeLocalDelayed)`)
```

Comment:
I'm pretty sure that an await on a non-promise value returns that value, no?

Reply:
the promise might resolve to `null` or `undefined` but the error message is wrong then, that's true